### PR TITLE
Update SmallRye Health to 2.0

### LINKED ISSUE #2527: Update SmallRye Health Check to 2.0

### PR DESCRIPTION
**Pull request by antoinesd**
Fixes #2527 
Adding new servlet for `/health/live` and `/health/ready` and updating tests accordingly

### CODE REVIEW

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: This part is not right.

You will need 2 different build items for the liveness and readiness probes and make use of them here:
https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java#L67
and here:
https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java#L118

/cc @geoand

**Review by gsmet — CHANGES_REQUESTED** [review]
The interactions with the Kubernetes extensions are not right. See my comment inline.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @gsmet :+1: ! Indeed, when we added the initial Kubernetes Health support, there was only one thing to use. Now that this has changed, we should have separate build items that feed the appropriate properties.

@antoinesd If you don't feel like dealing with the Kubernetes stuff in this PR, you can remove those changes and I can deal with them in a subsequent PR

Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Ok will add. But should I revert what I did here ?

File: bom/pom.xml
Comment (mkouba) [inline bom/pom.xml]: `smallrye-health-2.0`?? Why do we use version in the artifactId?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Yes, those lines would need to removed and a couple new build items added (KubernetesLivenessPathBuildItem and KubernetesReadinessPathBuildItem) I guess. 
Should we be keeping the "legancy" `KubernetesHealthPathBuildItem` or not?

Comment (mkouba) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I suggest to cooperate on this PR and not create a new one for the kubernetes stuff.

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: OK, sounds like a plan.

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @antoinesd Looking at the health stuff, it seems like the user will be able to select whethe they choose simple `health` or the new `readiness` and `liveliness`, so I suggest that we keep `KubernetesHealthPathBuildItem` build only feed its data into the generation process when the others have not been set. WDYT?

File: bom/pom.xml
Comment (antoinesd) [reply bom/pom.xml]: It's the new versioning that @kenfinnigan introduced [here](https://groups.google.com/d/msg/smallrye/xs5HFiTq6gk/90pShrJXBgAJ).

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @geoand to be sure, do you mean that even previous code should be removed?

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @antoinesd No no :). I am just saying that you should remove these lines you have added here add something like:

```
        healthPathItemProducer.produce(new KubernetesHealthPathBuildItem(health.healthPath));
        livenessPathItemProducer.produce(new KubernetesLivenessPathBuildItem(health.livenessPath));
        readinessPathItemProducer.produce(new KubernetesReadinessPathBuildItem(health.readinessPath));
```

where the `KubernetesLivenessPathBuildItem` and `KubernetesReadinessPathBuildItem` are new build items. 

The of course changes will be needed at: https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java#L121 and https://github.com/quarkusio/quarkus/blob/master/extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java#L67 to use these values (and also take into account if they are missing and only `health` is set)

Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I would have named them `KubernetesHealthLivenessPathBuildItem` and `KubernetesHealthReadinessPathBuildItem` with `Health` in the name.

@geoand what I'm wondering is how you will choose between the old one and the new ones?

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I am thinking that the legacy value will only be used if the new - specific value is not used.

I'm cool with the name you propose, whatever you think is best

Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: But you have default values for all of them. So they will always be defined AFAICS?

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Ah yes... Now I see what you are saying. In that case I guess we should just use the liveness and readiness stuff and ditch health entirely for Kubernetes. WDYT?

Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Yes, my thought exactly.

Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: OK, lets do that then

File: bom/pom.xml
Comment (kenfinnigan) [reply bom/pom.xml]: Right, it signifies the MP version it aligns with

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (kenfinnigan) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Not sure if this is possible, but is the `healthPath` the root for the `livenessPath` and `readinessPath`?

Or do we expect a user to be able to change `/health` to `/bubbles` and not impact the other two?

File: bom/pom.xml
Comment (mkouba) [reply bom/pom.xml]: Ok. I don't like it. But it's my fault that I missed the discussion. It's IMHO even more confusing if a GAV contains multiple versions.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (mkouba) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: +1 for having `healthPath` as a root for live and ready.

**Comment by gsmet** [comment]
@geoand @antoinesd I let you drive this puppy home when you have time.

**Comment by geoand** [comment]
@antoinesd ping me whenever you have time and if you need some assistance with this :)

**Comment by cescoffier** [comment]
Aby update on this PR?

**Comment by kenfinnigan** [comment]
@antoinesd what's the status on this? Do you need help?

**Comment by antoinesd** [comment]
Just pushed modifications. Sorry for the delay

File: extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java
Comment (geoand) [inline extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: I think we generally disallow `*`?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I would say that this part has been fixed now.

File: extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java
Comment (antoinesd) [reply extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: I think too. Strange that checkstyle didn't warn me on this during my local build

**Comment by geoand** [comment]
I also think the PR should be rebased onto master since we have now enabled some MP TCK checks (health among them)

**Comment by antoinesd** [comment]
@geoand rebase done

**Comment by geoand** [comment]
LGTM, @gsmet mind taking another look to ensure your comments have been addressed?

**Comment by geoand** [comment]
@antoinesd the TCK failed... does it  perhaps needs to be updated with for the new version of the smallrye health?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: This is not entirely true. It's the root path of the probes.

I don't think making it the root path AND the legacy path will work as, as soon as you define the root path, you end up enabling the legacy probe whereas you might not want it.

I think I would get rid of the legacy probe entirely. We are still in early stages of Quarkus, I think we can do this sort of changes.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I'm wondering if using `@ConfigItem(defaultValue = "${healthPath}/live")` would work?

If it's just the root path, I think it would be better.

File: extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java
Comment (gsmet) [inline extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: I think I would get rid of this one.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Either use them or remove them.

If we can use recursive resolution as I mentioned above, it would simplify all of this.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: You point to the exact same servlet for liveness and readiness. It feels weird?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I would remove this one.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a couple of comments. Could you take a look? Thanks!

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I think `rootPath` would be better.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I would drop the `Rel` part. Otherwise, you end up with: `quarkus.smallrye-health.liveness-rel-path`. Not very appealing.

If the other is `rootPath`, I think it's clear enough.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (geoand) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: For me personally, having a health root path that affects this value is confusing. I would either leave it the way @antoinesd has it, or we really want to make things clear and don't really mind about 100% backward compatibility at this point, I would drop the `healthPath` field completely.

Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: So my opinion is that we should drop exposing the servlet on the legacy `healthPath`.

So `healthPath` would be `rootPath`. I'm also not convinced it has any interest to have it instead of just `readinessPath` and `livenessPath`. Not sure if they plan to add more probes in MP Health. In this case, it could make sense.

Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @kenfinnigan you were the one asking for a root path. Do you think it has some strong value for the future?

**Comment by antoinesd** [comment]
@geoand not sure. I will update to last version and launch TCK locally. I wasn't aware that a profile was required to launch them

**Comment by geoand** [comment]
Yeah, the TCKs are a new addition and not enabled by default

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: If we do that Quarus won't support MP Health 2.0 specification since implementations must provide probe for `/health`. Do we really want that?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I didn't know it was possible to write this. Of course, it's far better to go that way

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: forgot them, sorry

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: probably an error

Comment (mkouba) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Not only it feels weird. It's imho wrong - this servlet always reports an overall health - all available checks. See https://github.com/smallrye/smallrye-health/blob/master/tck/src/main/java/io/smallrye/health/SmallRyeHealthServlet.java#L19.

Comment (mkouba) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: It should be probably `SmallRyeLivenessServlet` and `SmallRyeReadinessServlet`.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: Ah, OK. I wasn't aware of that. So, in this case, is it really legacy?

File: extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java
Comment (gsmet) [reply extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: OK, even if we keep the root probe, I think it does not make sense to have it here.

@geoand do you agree?

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (antoinesd) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: It was kept in 2.0 for backward compatibility, so yes it’s a kind of legacy

File: extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java
Comment (geoand) [reply extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: Give the fact that the other will always be set, yeah. How about we take it a step further and drop `kubernetesHealthPathBuildItem` completely?

Comment (gsmet) [reply extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java]: Yes, that's the obvious consequence :).

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (kenfinnigan) [reply extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: @gsmet my thought with a root path was that then it would be possible to change the path for everything under /health in one setting, instead of needing to change to paths

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: I think we should change it to say that it is the root path for the SmallRye Health servlet.

It being the path of the legacy servlet is a direct consequence of it and should be hidden, considered it's just a legacy behavior.

File: extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java
Comment (gsmet) [inline extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java]: As mentioned above, there's a good chance this one can entirely go away.

**Review by gsmet — CHANGES_REQUESTED** [review]
OK, I made a final review. Could you take a look?

I also think we should squash everything once it's done.

**Comment by antoinesd** [comment]
@gsmet last corrections and rebase done

**Comment by mkouba** [comment]
@gsmet Is it fixed? I think this could be a good candidate for the next release.

**Comment by gsmet** [comment]
@mkouba I just have one minor adjusment to do before merging it. Agreed it should be in the next release, it's the next PR I'm working on.

**Comment by gsmet** [comment]
I rebased, added a cleanup commit and I'm now waiting for the CI run before merging.

**Comment by geoand** [comment]
I have also tested the kubernetes / health check stuff and it works like a charm

**Comment by geoand** [comment]
@gsmet OK to merge?

**Review by gsmet — APPROVED** [review]

**Pull request merged**
